### PR TITLE
The big renaming

### PR DESCRIPTION
**Pull request by fingolfin**
Companion to https://github.com/Nemocas/AbstractAlgebra.jl/pull/1267, https://github.com/thofma/Hecke.jl/pull/977, https://github.com/oscar-system/Oscar.jl/pull/1972

This is WIP on the big type renaming, see also <https://hackmd.io/0hrBHIUrRZG23xiZZ0NCEA?both> for our notes (these are partially outdated and misleading, we are working on that).

One issue with this PR is that we rename
- `FqDefaultPolyRing` => `FqPolyRing`
- `FqPolyRing` => `FqPolyRepPolyRing`

So the name `FqPolyRing` is both an old and a new type name, albeit with different meanings, which is unfortunate.

Actually, this could be avoided because `FqPolyRing` is not the final name we want anyway... Right now we do these renamings:
- `FqDefaultFiniteField` => `FqField`
- `FqDefaultMatSpace` => `FqMatrixSpace`
- `FqDefaultMPolyRing` => `FqMPolyRing`
- `FqDefaultPolyRing` => `FqPolyRing`
but we really would prefer to just call this `FinField`, `FinFieldMatrixSpace`, etc. (or perhaps `FiniteField`, ...?) but `FinField` is already an abstract type in AbstractAlgebra... We could rename that, or just re-use the name (by not importing `FinField` from AA into Nemo), but all of these involve further discussions and decisions which is why I did not yet attempt this here.

Also missing are the various SeriesRings/SeriesElem. I'll work on those tomorrow

And of course there are more things to be done in AA, Oscar, Hecke

TODOs (either in this PR or a follow up):
- [ ] also rename `_series` types
- [ ] consider updating filenames (such as `fmpz.jl`), OR if we keep them, ensure we reference them correctly in code and docs
- [ ] update documentation (e.g. `docs/src/developer/introduction.md:58`) with a table translating old "Flint names" like `fmpz` to new "Nemo names" like `ZZRingElem`
- [ ] add docstrings to all types that
      - explain what the type represents
      - mentions Flint names, if applicable
      - gives an example how to create an instance (ideally as a `jldoctest`)
- ...

### CODE REVIEW

**Comment by fingolfin** [comment]
Also, we wanted to rename `arb` to `RealFieldElem` and more arb/acb renaming, but then discovered that there is already `RealElem` which seems to be almost identical to `arb` except for the parents, and so maybe we need some further discussion, perhaps with @thofma ?

File: src/flint/fmpq_poly.jl [hunk 1]
```diff
@@ -1,26 +1,26 @@
 ###############################################################################
 #
-#   fmpq_poly.jl : Flint polynomials over fmpq
+#   QQPolyRingElem.jl : Flint polynomials over QQFieldElem
```
Comment (lgoe-ac) [inline src/flint/fmpq_poly.jl:3]: In case you didn't notice. Currently, the file names do not match the comments. So this seems to be a renaming artifact

Comment (fingolfin) [reply src/flint/fmpq_poly.jl:3]: Yes, indeed, thank you for pointing it out. We will discuss today if we also want to rename files or not, and then I'll resolve this depending on the outcome.

**Comment by fingolfin** [comment]
The failure in OscarCI is to be expected (a printing test fails, and an array `fmpz[]` is now of course being printed differently).

With Julia nightly it fails with this, which looks as if it is a separate issue:
> WARNING: both Nemo and Base export "tanpi"; uses of it in module Main must be qualified

**Comment by thofma** [comment]
https://github.com/JuliaLang/julia/pull/48575 :-(

**Comment by fingolfin** [comment]
> I merged #1367 yesterday, because it is related.

Thank you, I hadn't noticed, and will read up there now.

> 
> To quote from [#1367 (comment)](https://github.com/Nemocas/Nemo.jl/pull/1367#issuecomment-1374791490)
> 
> > I don't want to hear any discussions about the names. If they need to be adjusted, it can be done during the "big rename"

Oh sure -- I didn't mean to complain  about the names there (or anything, for that matter), it was just a bit of confusion yesterday when we discovered `RealElem` vs. `arb` and where unsure how to continue.

But now this seems clearer:

> 
> Ideally a user will ever see an `arb/acb` again, so I would leave them as they are now. Rename the `RealElem` etc. as much as you want. I am also happy to discuss this.

So basically we use `RealElem` (possibly renaming it to `RealFieldElem`, etc., etc and ignore `arb` and `acb` and friends. And then in Oscar.jl, I think the only places referencing them are in the serialization code, so I guess that code should be changed to use `RealElem` etc. (And for Hecke, I am sure you already have a plan, so we'll not touch it there)

**Comment by thofma** [comment]
Yes, that sounds good.

File: src/flint/fmpq_poly.jl (hunk 1 above)
Comment (fingolfin) [reply src/flint/fmpq_poly.jl:3]: I hopefully fixed all headers in `src/flint/*.jl`

**Comment by fingolfin** [comment]
I've now renamed `RealElem` to `RealFieldElem` and `ComplexElem` to `ComplexFieldElem`.

Alas, there is already a type `RealFieldElem` in Hecke, for a completely different purpose, so I'll have to rename *that* first now ...

**Comment by thofma** [comment]
there is a reason it is *big* renaming and not just renaming :)

**Comment by thofma** [comment]
Is squash fine?

**Comment by fingolfin** [comment]
ARGH I accidentally rebased instead of a merge. I'll fix it

**Pull request merged**